clients/java/src/main/java/io/zeebe/client/impl/RetriableStreamingFutureImpl.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.client.impl;
+
+import io.grpc.stub.StreamObserver;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+public class RetriableStreamingFutureImpl<C, B> extends ZeebeStreamingClientFutureImpl<C, B> {
+
+  private final Predicate<Throwable> retryPredicate;
+  private final Consumer<StreamObserver<B>> retryAction;
+
+  public RetriableStreamingFutureImpl(
+      C clientResponse,
+      Consumer<B> collector,
+      Predicate<Throwable> retryPredicate,
+      Consumer<StreamObserver<B>> retryAction) {
+    super(clientResponse, collector);
+    this.retryPredicate = retryPredicate;
+    this.retryAction = retryAction;
+  }
+
+  @Override
+  public void onError(Throwable throwable) {
+    if (retryPredicate.test(throwable)) {
```

Comment:
Considering we always need both `retryPredicate` and `retryAction`, could we have a specific interface, e.g. `RetryStrategy` which implements both the predicate and the action?

```java
public interface RetryStrategy {
  boolean shouldRetryRequest(Throwable error);
  void retryRequest(StreamObserver<?> request);
}
```

I'm just not sure of the value of splitting them into two different components, since they seem softly coupled. What do you think?

Reply:
I understand what you mean and looking it at it from the RetriableFuture's perspective, I'd agree it'd be better. But you'd have to implement the interface in each command because in the end the predicate comes from the provider and the retryRequest from the command itself. You'd have an abstraction just for the retriable future and I'm not sure that's very useful

Reply:
But do you think it's still valuable?